students/paul_c/session05/mailroom.py

```diff
@@ -0,0 +1,156 @@
+#!/usr/local/bin/python3
+
+import sys
+import time
+
+donors = {
+    "Anne Ant": [1.00],
+    "Bonnie Bug": [20.00, 10.00],
+    "Chuck Cat": [2.00, 2.00, 2.00],
+    "Donna Dog": [5000.00, 2500.00, 1.50],
+    "Edna Ent": [.30, .50, .10]
+}
+
+email = {
+    "greeting": "\nHello {}\n\n",
+    "body": "We would like to thank you for your generous donation of ${}.\n\n",
+    "closing": "Best Regards,\n",
+    "signature": "The Foundation\n\n"
+}
+
+
+def send_all():
+    """
+    Function to send a thank you letter to all donors.
+    This function sums total donations and writes letters to timestamped files.
+    """
+    for key, value in donors.items():
+        timestamp = time.strftime("%b %d %Y %H:%M:%S")
+        filename = f"{timestamp}-{key}.txt"
+        file_content = '{greeting}' '{body}' '{closing}' '{signature}'.format(**email).format(key, sum(value))
+        files = open(filename, "w")
+        files.write(file_content)
+        files.close()
+        print(str(files) + "  ---File created.\n")
+
+
+def create_report():
+    """
+    This function prints out a report with the following parameters:
+    Donor Name, Total Given, Number of Gifts, Average Gift Amount
+    """
+    header = ("Donor Name", "| Total Given", "| Num Gifts", "| Average Gift")
+    row = " ".join(["{:20s} {:>20s} {:>20s} {:>20s}"]).format(*header)
+    header_length = len(row)
+    print("\n" + row)
+    print("=" * header_length)
+    for key, value in sorted(donors.items()):
```

Comment:
Try converting this one to use list comprehension. To make it a bit simpler to read, I'd use f-strings. Here's a hint:

```
    result = [f"{key:20s} ... for key, value in ...)]
    for donor in result:
        print(donor)


```-Luis

Reply:
Thank you for the hint. I was able to turn that portion of code into the following.

donor_row = [f"{key:20s} {sum(value):>20.2f} {len(value):>20.2f} {sum(value)/len(value):>20.2f}" for key, value in sorted(donors.items())]